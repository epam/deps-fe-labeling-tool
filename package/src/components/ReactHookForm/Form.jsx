import React from 'react'
import PropTypes from 'prop-types'
import isRequiredIf from 'react-proptype-conditional-require'
import { childrenShape } from '@/utils/propTypes'

const FormValidationMode = {
  ON_CHANGE: 'onChange',
  ON_BLUR: 'onBlur',
  ON_SUBMIT: 'onSubmit',
  ON_TOUCHED: 'onTouched',
  ALL: 'all'
}

const Form = ({
  children,
  onSubmit,
  handleSubmit,
  ...rest
}) => (
  <form
    onSubmit={handleSubmit?.(onSubmit)}
    {...rest}
  >
    {children}
  </form>
)

Form.propTypes = {
  children: childrenShape.isRequired,
  onSubmit: PropTypes.func,
  handleSubmit: isRequiredIf(
    PropTypes.func, (props) => props.onSubmit
  )
}

export {
  Form,
  FormValidationMode
}
