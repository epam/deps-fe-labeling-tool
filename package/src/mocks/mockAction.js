const mockAction = {
  type: 'ANY_ACTION'
}

export {
  mockAction
}
