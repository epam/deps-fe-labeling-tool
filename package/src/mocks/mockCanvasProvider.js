const mockCanvasProvider = {
  withCanvas: (WrappedComponent) => ({
    WrappedComponent
  })
}

export {
  mockCanvasProvider
}
