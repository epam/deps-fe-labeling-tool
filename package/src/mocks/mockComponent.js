const mockComponent = (componentName, namedExport = true, fn = () => { }) => {
  Object.defineProperty(fn, 'name', { value: componentName })
  return !namedExport ? fn : { [componentName]: fn }
}

export {
  mockComponent
}
