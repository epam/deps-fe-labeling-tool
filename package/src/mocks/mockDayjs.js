
const formats = {
  L: 'MM/DD/YYYY',
  LT: 'h:mm A',
  default: 'MM/DD/YYYY, HH:mm'
}

const MOCKED_DATE = '2023-10-03T12:14:02'

const mockDayjs = (dateTimeStr = MOCKED_DATE) => ({
  __esModule: true,
  default: (() => {
    function dayjs (date = dateTimeStr) {
      return {
        valueOf: () => new Date(date).getTime(),
        format: () => date,
        isValid: () => true,
        diff: () => -100500,
        add: () => dayjs(date),
        startOf: () => dayjs(date),
        endOf: () => dayjs(date),
        subtract: () => dayjs(date),
        longDateFormat: (format = 'default') => formats[format],
        localeData: () => dayjs(date),
        toISOString: () => date
      }
    }
    dayjs.locale = () => {}
    dayjs.extend = () => {}
    dayjs.utc = () => dayjs(dateTimeStr)
    return dayjs
  })()
})

export {
  mockDayjs
}
