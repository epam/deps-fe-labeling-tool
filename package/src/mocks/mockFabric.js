class MockPoint {
  static instance

  x

  y

  constructor (x, y) {
    this.x = x
    this.y = y
    MockPoint.instance = this
  }

  static getLastCreatedInstance () {
    return MockPoint.instance
  }
}

class MockCanvas {
  ref

  width = 1366

  height = 768

  zoom = 1

  backgroundColor = '#ffffff'

  backgroundImage = {}

  viewportTransform = [0, 0, 0, 0, 0, 0]

  selection = true

  listeners = {}

  static instance

  constructor (ref = undefined) {
    this.ref = ref
    MockCanvas.instance = this
  }

  static getLastCreatedInstance () {
    return MockCanvas.instance
  }

  on = jest.fn((eventName, callback) => {
    this.listeners[eventName] = this.listeners[eventName] || []
    this.listeners[eventName].push(callback)
  })

  off = jest.fn((eventName, callback) => {
    this.listeners[eventName] = this.listeners[eventName].filter((registeredCallback) => registeredCallback !== callback)
  })

  setViewportTransform = jest.fn((vpt) => {
    this.viewportTransform = vpt
  })

  getWidth = jest.fn(() => this.width)

  setWidth = jest.fn((width) => {
    this.width = width
  })

  getHeight = jest.fn(() => this.height)

  setHeight = jest.fn((height) => {
    this.height = height
  })

  getZoom = jest.fn(() => this.zoom)

  setZoom = jest.fn((zoom) => {
    this.zoom = zoom
  })

  getCenter = jest.fn(() => ({
    left: this.width / 2,
    top: this.height / 2
  }))

  setBackgroundColor = jest.fn((backgroundColor) => {
    this.backgroundColor = backgroundColor
  })

  setBackgroundImage = jest.fn((backgroundImage) => {
    this.backgroundImage = backgroundImage
  })

  zoomToPoint = jest.fn((point, zoom) => {
    this.viewportTransform[4] = point.x
    this.viewportTransform[5] = point.y
    this.zoom = zoom
  })

  getObjects = jest.fn(() => [{
    data: {
      uid: '1'
    },
    ungroupOnCanvas: jest.fn(() => ({ getObjects: jest.fn(() => ['testEl']) })),
    getObjects: jest.fn(() => ['testEl'])
  }])

  requestRenderAll = jest.fn()

  renderAll = jest.fn()

  add = jest.fn()

  remove = jest.fn()

  fxRemove = jest.fn()

  findTarget = jest.fn()

  discardActiveObject = jest.fn()

  setCursor = jest.fn()

  add = jest.fn()

  sendToBack = jest.fn()

  setActiveObject = jest.fn()

  getActiveObject = jest.fn()

  getActiveObjects = jest.fn(() => [])

  fire = jest.fn()

  getPointer = jest.fn()
}

class MockImage {
  width = 0

  height = 0

  src = '';

  constructor (image) {
    this.src = (image && image.src) || 'MOCK_SRC'
    this.width = 288
    this.height = 512
  }

  static fromURL = jest.fn().mockImplementation((src, callback) => {
    const image = new MockImage()
    image.src = src
    callback(image)
  })

  getSrc () {
    return this.src
  }
}

class MockActiveSelection {
  set = jest.fn()

  on = jest.fn()
}

class MockGroup {

}

class MockText {

}

const MockTextbox = jest.fn(() => ({
  set: jest.fn(),

  setCoords: jest.fn()
}))

class MockRect {
  on = jest.fn()

  set = jest.fn()

  setCoords = jest.fn(() => this)

  getScaledWidth = jest.fn()

  getScaledHeight = jest.fn()

  setControlsVisibility = jest.fn()

  left = 10

  top = 10

  height = 10

  width = 10

  selectable = true

  name = 'mockRect'

  content = 'mockRect'

  data = {
    uid: 0
  }
}

class MockCircle {
  set = jest.fn()
}

const mockFabric = {
  fabric: {
    Text: MockText,
    Image: MockImage,
    Canvas: MockCanvas,
    Point: MockPoint,
    Group: MockGroup,
    Textbox: MockTextbox,
    Rect: MockRect,
    Circle: MockCircle,
    ActiveSelection: MockActiveSelection
  }
}

class MockTable {
  set = jest.fn()

  on = jest.fn()
}

class MockCell {
  set = jest.fn()

  on = jest.fn()
}

class MockBorder {
  set = jest.fn()
}

class MockSplitProjection {
  set = jest.fn()
}

class MockFabricArrow {
  set = jest.fn()

  setCoords = jest.fn()
}

export {
  MockText,
  MockPoint,
  MockCanvas,
  MockImage,
  MockRect,
  mockFabric,
  MockCell,
  MockTable,
  MockBorder,
  MockSplitProjection,
  MockActiveSelection,
  MockFabricArrow
}
