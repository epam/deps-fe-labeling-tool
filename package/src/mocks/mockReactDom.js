const mockReactDom = ({
  render: jest.fn(),
  unmountComponentAtNode: jest.fn()
})

export {
  mockReactDom
}
