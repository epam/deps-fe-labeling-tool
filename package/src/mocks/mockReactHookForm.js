import { FieldType } from '@/enums/FieldType'

const mockReactHookForm = ({
  ...jest.requireActual('react-hook-form'),
  useForm: jest.fn(() => ({
    handleSubmit: jest.fn(),
    getValues: jest.fn(),
    formState: {}
  })),
  useController: jest.fn(() => ({
    field: {},
    fieldState: {}
  })),
  useFormContext: jest.fn(() => ({
    control: {},
    setValue: jest.fn()
  })),
  useWatch: jest.fn(() => FieldType.STRING)
})

export {
  mockReactHookForm
}
