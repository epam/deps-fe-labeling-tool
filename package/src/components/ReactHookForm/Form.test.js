import { mockReactHookForm } from '@/mocks/mockReactHookForm'
import React from 'react'
import { shallow } from 'enzyme'
import { Form } from './Form'

jest.mock('react-hook-form', () => mockReactHookForm)

describe('Component: Form', () => {
  let defaultProps
  let wrapper

  beforeEach(() => {
    defaultProps = {
      children: <input />,
      onSubmit: jest.fn(),
      handleSubmit: jest.fn()
    }

    wrapper = shallow(<Form {...defaultProps} />)
  })

  it('should render correct Form layout', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should call handleSubmit with correct props when form is submitted', () => {
    wrapper.find('form').simulate('submit')
    expect(defaultProps.handleSubmit).nthCalledWith(1, defaultProps.onSubmit)
  })
})
