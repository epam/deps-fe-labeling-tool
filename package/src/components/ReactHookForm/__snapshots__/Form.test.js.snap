// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Form should render correct Form layout 1`] = `
<form>
  <input />
</form>
`;
